contrib/src/main/java/com/twelvemonkeys/contrib/tiff/TIFFUtilities.java

```diff
@@ -340,66 +343,210 @@ private long write(ImageOutputStream outputStream, TIFFWriter tiffWriter) throws
+                        throw new IllegalArgumentException("Old-style-JPEG with multiple strips are only supported, if first strip contains SOS");
+                    }
+
+
+                    newOffsets = new int[offsets.length];
+                    int[] newByteCounts = new int[byteCounts.length];
+                    for (int i = 0; i < offsets.length; i++) {
+                        newOffsets[i] = (int) outputStream.getStreamPosition();
+                        outputStream.write(jpegInterchangeData);
+                        stream.seek(offsets[i]);
+
+                        byte[] buffer = new byte[(int) byteCounts[i]];
+                        newByteCounts[i] = (int) (jpegInterchangeData.length + byteCounts[i]);
+                        stream.readFully(buffer);
+                        if (buffer[0] != 0xff && buffer[1] != 0xda) {
+                            outputStream.write(SOSMarker);
+                            newByteCounts[i] += SOSMarker.length;
+                        }
+                        outputStream.write(buffer);
+                    }
+
+                    newIFD.remove(stripOffsetsEntry);
+                    newIFD.add(new TIFFEntry(useTiles ? TIFF.TAG_TILE_OFFSETS : TIFF.TAG_STRIP_OFFSETS, newOffsets));
+                    newIFD.remove(stripByteCountsEntry);
+                    newIFD.add(new TIFFEntry(useTiles ? TIFF.TAG_TILE_BYTE_COUNTS : TIFF.TAG_STRIP_BYTE_COUNTS, newByteCounts));
+
+                    newIFD.remove(oldJpegData);
+                    newIFD.remove(oldJpegDataLength);
+                    rearrangedByteStrips = true;
+                }
+            }
+
+
+            if (!rearrangedByteStrips && stripOffsetsEntry != null && stripByteCountsEntry != null) {
                 newOffsets = writeData(offsets, byteCounts, outputStream);
 
                 newIFD.remove(stripOffsetsEntry);
-                newIFD.add(new TIFFEntry(TIFF.TAG_STRIP_OFFSETS, newOffsets));
+                newIFD.add(new TIFFEntry(useTiles ? TIFF.TAG_TILE_OFFSETS : TIFF.TAG_STRIP_OFFSETS, newOffsets));
             }
 
-            Entry oldJpegData = IFD.getEntryById(TIFF.TAG_JPEG_INTERCHANGE_FORMAT);
-            Entry oldJpegDataLength = IFD.getEntryById(TIFF.TAG_JPEG_INTERCHANGE_FORMAT_LENGTH);
-            if (oldJpegData != null && oldJpegData.valueCount() > 0 && oldJpegDataLength != null && oldJpegDataLength.valueCount() > 0) {
-                if (!Arrays.equals(getValueAsLongArray(oldJpegData), offsets) || !Arrays.equals(getValueAsLongArray(oldJpegDataLength), byteCounts)) {
-                    // data already written from TIFF.TAG_STRIP_OFFSETS
-                    offsets = getValueAsLongArray(oldJpegData);
-                    byteCounts = getValueAsLongArray(oldJpegDataLength);
-                    newOffsets = writeData(offsets, byteCounts, outputStream);
+            Validate.isTrue(oldJpegData == null || !newIFD.contains(oldJpegData), "Failed to transform old-style JPEG");
```

Comment:
Are you sure you want to throw `IllegalArgumentException` here? Shouldn't this be an `IOException` (or subclass), if we failed to transcode? There's nothing the programmer can reasonably do to prevent this.